lib/charms/opensearch/v0/opensearch_secrets.py

```diff
@@ -0,0 +1,300 @@
+        components = label.split(self.LABEL_SEPARATOR)
+
+        unit_id = None
+        key = None
+        if len(components) == 3:
+            key = components[2]
+        elif len(components) == 4:
+            unit_id = int(components[2])
+            key = components[3]
+
+        if components[1] == Scope.APP.val:
+            scope = Scope.APP
+        elif components[1] == Scope.UNIT.val:
+            scope = Scope.UNIT
+
+        return {
+            "application_name": components[0],
+            "scope": scope,
+            "unit_id": unit_id,
+            "key": key,
+        }
+
+    @staticmethod
+    def _safe_obj_data(indict: Dict) -> Dict:
+        return {key: str(val) for key, val in indict.items() if val is not None}
+
+    def _get_juju_secret(self, scope: Scope, key: str) -> Optional[Secret]:
+        label = self.label(scope, key)
+
+        cached_secret_meta = self.cached_secrets.get_meta(scope, label)
+        if cached_secret_meta:
+            return cached_secret_meta
+
+        try:
+            secret = self._charm.model.get_secret(label=label)
+        except SecretNotFoundError:
+            return None
+
+        self.cached_secrets.set_meta(scope, label, secret)
+        return secret
+
+    def _get_juju_secret_content(self, scope: Scope, key: str) -> Optional[Dict]:
+        cached_secret_content = self.cached_secrets.get_content(scope, self.label(scope, key))
+        if cached_secret_content:
+            return cached_secret_content
+
+        secret = self._get_juju_secret(scope, key)
+        if not secret:
+            return None
+
+        content = secret.get_content()
+        self.cached_secrets.set_content(scope, self.label(scope, key), content=content)
+        return content
+
+    def _save_secret_id(self, scope_obj: Union[Application, Unit], label: str, secret_id: str):
+        """Keeping a reference of the secret's ID just for sure."""
+        relation = self.model.get_relation(self.relation_name)
+        if relation:
+            relation.data[scope_obj][label] = secret_id
+        else:
+            logger.error("Couldn't save secret ID %s -- %s", label, secret_id)
+
+    def _add_juju_secret(self, scope: Scope, key: str, value: Dict[str, str]) -> Optional[Secret]:
+        safe_value = self._safe_obj_data(value)
+
+        if not safe_value:
+            return None
+
+        scope_obj = None
+        if scope == Scope.APP:
+            scope_obj = self._charm.app
+        if scope == Scope.UNIT:
+            scope_obj = self._charm.unit
+
+        label = self.label(scope, key)
+        try:
+            secret = scope_obj.add_secret(safe_value, label=label)
+        except ValueError:
+            logging.error("Secert %s:%s couldn't be added", str(scope.val), str(key))
+            return None
+
+        self.cached_secrets.update(scope, label, secret, safe_value)
+        self._save_secret_id(scope_obj, label, secret.id)
+        return secret
```

Comment:
`ValueError` typically happens when you want to insert an empty string, perhaps also when value is of a wrong type (i.e. not string). If needed I can double check the latter.

All in all: in the current code (according to my best knowledge) this is not supposed to happen. With future developments it may be a very valid use case to pass a wrong value that would be refused by Juju Secrets to be saved.

Reply:
https://github.com/canonical/opensearch-operator/pull/102#discussion_r1304026606 applies here too